src/main/webapp/view/discussion/discussion-manager.js

```diff
@@ -24,57 +24,60 @@ export default class DiscussionManager {
   static #PARAM_TIMESTAMP = 'timestamp';
   static #PARAM_TYPE = 'type';
   #lecture;
+  #displayedComments;
 
   /**
    * Creates a `DiscussionManager` to manage posting and fetching comments for
    * the `lecture`.
    */
   constructor(lecture) {
     this.#lecture = lecture;
+    this.#displayedComments = new Map();
   }
 
   /**
-   * Organizes comments into threads with nested replies.
-   *
-   * <p>All comments are sent from the servlet without any structure, so the
-   * client needs to organize them before displaying.
+   * Adds new comments to threads with nested replies by setting comment's
+   * parent and replies fields. Only processes and returns comments with ids
+   * that haven't already been seen. This is a private method and should only
+   * be called by `DiscussionManager`.
    */
-  structureComments(allComments) {
-    const commentIds = {};
+  processNewComments(allComments) {
+    const newComments = [];
     for (const comment of allComments) {
+      const id = comment.commentKey.id;
+      if (this.#displayedComments.has(id)) {
+        // This comment is already displayed.
+        continue;
+      }
       comment.replies = [];
-      commentIds[comment.commentKey.id] = comment;
+      newComments.push(comment);
+      this.#displayedComments.set(id, comment);
     }
 
-    const rootComments = [];
-    for (const comment of allComments) {
+    for (const comment of newComments) {
```

Comment:
Is there a reason this for loop is separate from the other? Do replies sometimes get processed before their parent comment?

Reply:
Yeah.. I believe the order that comments are sent is random, so I can't assume that parent comments will be processed before their replies. Therefore I need to do some minor logic (storing in map, adding `.replies = []`, etc.) on everything before actually organizing them into threads

Reply:
Well, not random... But they're not reliably sorted in any meaningful way

Reply:
Gotcha, can you add a comment explaining it a bit? (and for the other for loop pair as well)

Reply:
Added!